SU2_CFD/include/numerics_simd/CNumericsSIMD.hpp

```diff
@@ -0,0 +1,711 @@
+FORCEINLINE VectorDbl<nDim+2> inviscidProjFlux(const CCompressiblePrimitives<nDim>& V,
+                                               const CCompressibleConservatives<nDim>& U,
+                                               const VectorDbl<nDim>& normal) {
+  auto mdot = dot(U.momentum(), normal);
+  VectorDbl<nDim+2> flux;
+  flux(0) = mdot;
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    flux(iDim+1) = mdot*V.velocity(iDim) + normal(iDim)*V.pressure();
+  }
+  flux(nDim+1) = mdot*V.enthalpy();
+  return flux;
+}
+
+template<size_t nDim, class RandomIterator>
+FORCEINLINE MatrixDbl<nDim+2> inviscidProjJac(Double gamma, RandomIterator velocity,
+                                              Double energy, const VectorDbl<nDim>& normal,
+                                              Double scale) {
+  MatrixDbl<nDim+2> jac;
+
+  auto projVel = dot(velocity, normal);
+  auto gamma_m_1 = gamma-1;
+  auto phi = 0.5*gamma_m_1*squaredNorm<nDim>(velocity);
+  auto a1 = gamma*energy - phi;
+
+  jac(0,0) = 0.0;
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    jac(0,iDim+1) = scale * normal(iDim);
+  }
+  jac(0,nDim+1) = 0.0;
+
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    jac(iDim+1,0) = scale * (normal(iDim)*phi - velocity[iDim]*projVel);
+    for (size_t jDim = 0; jDim < nDim; ++jDim) {
+      jac(iDim+1,jDim+1) = scale * (normal(jDim)*velocity[iDim] - gamma_m_1*normal(iDim)*velocity[jDim]);
+    }
+    jac(iDim+1,iDim+1) += scale * projVel;
+    jac(iDim+1,nDim+1) = scale * gamma_m_1 * normal(iDim);
+  }
+
+  jac(nDim+1,0) = scale * projVel * (phi-a1);
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    jac(nDim+1,iDim+1) = scale * (normal(iDim)*a1 - gamma_m_1*velocity[iDim]*projVel);
+  }
+  jac(nDim+1,nDim+1) = scale * gamma * projVel;
+
+  return jac;
+}
+
+template<size_t nVar>
+FORCEINLINE void updateLinearSystem(Int iEdge,
+                                    Int iPoint,
+                                    Int jPoint,
+                                    bool implicit,
+                                    UpdateType updateType,
+                                    const VectorDbl<nVar>& flux,
+                                    const MatrixDbl<nVar>& jac_i,
+                                    const MatrixDbl<nVar>& jac_j,
+                                    CSysVector<su2double>& vector,
+                                    CSysMatrix<su2mixedfloat>& matrix) {
+  if (updateType == UpdateType::COLORING) {
+    vector.AddBlock(iPoint, flux);
+    vector.SubtractBlock(jPoint, flux);
+    if(implicit) matrix.UpdateBlocks(iEdge, iPoint, jPoint, jac_i, jac_j);
+  }
+  else {
+    vector.SetBlock(iEdge, flux);
+    if (implicit) matrix.SetBlocks(iEdge, jac_i, jac_j);
+  }
+}
+
+/*!
+ * \class CRoeBase
+ * \brief Base class for Roe schemes, derived classes implement
+ *        the dissipation term in a "finalizeResidual" method.
+ */
+template<class Derived, size_t NDIM>
+class CRoeBase : public CNumericsSIMD {
```

Comment:
For example the base class for Roe schemes is templated on its "future" derived classes.